L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,504 @@
+  // @param tls_version: The maximum TLS version.
+  void set_max_tls_version(grpc_tls_version tls_version);
+
+  // WARNING: EXPERT USE ONLY. MISUSE CAN LEAD TO SIGNIFICANT SECURITY DEGRADATION.
+  //
+  // Sets a custom chain builder implementation that replaces the default chain
+  // building from the underlying SSL library. Fully replacing and implementing
+  // chain building is a complex task and has dangerous security implications if
+  // done wrong, thus this API is inteded for expert use only.
+  void set_custom_chain_builder(std::shared_ptr<CustomChainBuilder>
+  chain_builder);
+}
+
+// Server-specific options for configuring TLS.
+class TlsServerCredentialsOptions final : public TlsCredentialsOptions {
+ public:
+  // A certificate provider that provides identity credentials is required,
+  // because a server must always present identity credentials during any TLS
+  // handshake. The certificate provider may optionally provide root
+  // certificates, in case the server requests client certificates.
+  explicit TlsServerCredentialsOptions(
+      std::shared_ptr<CertificateProviderInterface> certificate_provider)
+      : TlsCredentialsOptions() {
+    set_certificate_provider(certificate_provider);
+  }
+
+
+  // Sets requirements for if client certificates are requested, if they
+  // are required, and if the client certificate must be trusted. The
+  // default is GRPC_SSL_DONT_REQUEST_CLIENT_CERTIFICATE, which represents
+  // normal TLS.
+  void set_cert_request_type(
+      grpc_ssl_client_certificate_request_type cert_request_type);
+}
+
+// Client-specific options for configuring TLS.
+//
+// A client may optionally set a certificate provider.
+// If there is no certificate provider, the system default root certificates will be used to
+// verify server certificates. 
+// If a certificate provider is set and it provides root certifices that root
+// will be used.
+// If a certificate provider is set and it provides identity credentials those
+// identity credentials will be used.
+class TlsChannelCredentialsOptions final : public TlsCredentialsOptions {
+ public:
+  // Sets the decision of whether to do a crypto check on the server certificates.
+  // The default is true.
+  void set_verify_server_certificates(bool verify_server_certs);
+}
+
+// Builds a ChannelCredentials instance that establishes TLS connections in the manner specified by options.
+std::shared_ptr<ChannelCredentials> TlsCredentials(
+    const TlsChannelCredentialsOptions& options);
+}
+
+// Builds a ServerCredentials instance that establishes TLS connections in the manner specified by options.
+std::shared_ptr<ServerCredentials> TlsServerCredentials(
+    const TlsServerCredentialsOptions& options);
+```
+
+To make migration from the `SslCredentials` API to the `TlsCredentials` API easier, we will also have a constructor for `TlsChannelCredentialsOptions` that accepts the `SslCredentialsOptions`
```

Comment:
From review, consider having a method that returns the TLS options from the SSL options, rather than having a constructor.

Reply:
I think the comment from the review was something different than what made it into the text: IIRC, the idea from the review was to have a standalone "convert SSL to TLS options" API, rather than making it a member method of any class or struct. This way we avoid leaking details of the TLS class into the SSL class or vice-versa.

Reply:
Changed text to make it clearer